### PR TITLE
fix: resolve 422 error during column deletion

### PR DESCRIPTION
**Pull request by atifxcode**
## Description
While testing the transformation pipeline locally, I encountered a 422 Unprocessable Entity error during column deletion. The AddOrDeleteColumn schema was strictly requiring a name string, which is redundant for delCol.

This PR refactors the schema to allow optional names and implements conditional validation in the backend to ensure data integrity for additions.

Fixes #(issue number)
Found and resolved during local testing

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?

Describe the tests you ran to verify your changes.

- [ ] Existing tests pass
- [ ] New tests added
- [x] Manual testing:
  - Verified via DevTools that `delCol` POST requests now return 200 OK.
  - Confirmed column removal works in the UI.
  - Verified `addCol` still works (and still requires a name).

## Screenshots (if applicable)

Before : 
https://github.com/user-attachments/assets/57c89ab2-ad16-4565-8888-0c16c289be9b

After :
https://github.com/user-attachments/assets/9910616b-4749-4b9b-b997-33f56ce0715f

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [ ] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally : All the test (frontend and backend) both passed successfully locally.

## Changes Made
Backend:
Updated AddOrDeleteColumn schema to make the name field optional to support deletion.

## Frontend:
Updated the transformProject call in Table.jsx to remove the redundant name parameter during column deletion.

## Modified files
app/schemas.py — Schema refactor.
app/api/endpoints/transformations.py — Backend logic & validation.
src/Components/Table.jsx — Frontend API integration cleanup.

### CODE REVIEW

**Review by ivantha — CHANGES_REQUESTED** [review]
Please tag a valid issue ID. Also the frontend now sends name for delCol when it's not needed. This isn't a bug, but it creates a slight inconsistency. The whole point of the backend change was to make name optional for deletion, yet the frontend now includes it anyway. If someone reads the frontend code, they might think name is required for column deletion

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR correctly identifies and fixes the root cause of the 422 error (the schema requiring `name` for all column operations, including deletions that don't need one), and the backend conditional validation approach is reasonable. However, there is a likely **off-by-one bug** in `Table.jsx` where `columns[index]` retrieves the name of the *wrong* column (the one after the target), which means every deletion could silently operate on an inconsistent name/index pair. Additionally, the PR description directly contradicts the code change (it says "remove the redundant name parameter" but the code *adds* it), there is no automated test coverage for the fix, the `delCol` handler has no guard against a missing `col_params` payload, and empty-string column names bypass the `addCol` validation. The schema design also warrants a discussion about splitting `AddOrDeleteColumn` into two purpose-specific models for cleaner contracts.

File: dataloom-backend/app/api/endpoints/transformations.py [hunk 1]
```diff
@@ -57,8 +57,8 @@ def _handle_basic_transform(df, transformation_input, project, db, project_id):
         return ts.delete_row(df, transformation_input.row_params.index), True
 
     elif op == 'addCol':
-        if not transformation_input.col_params:
-            raise HTTPException(status_code=400, detail="Column parameters required")
+        if not transformation_input.col_params or transformation_input.col_params.name is None:
+            raise HTTPException(status_code=400, detail="Column index and name required for addCol")
```
Comment (ivantha) [inline dataloom-backend/app/api/endpoints/transformations.py]: **[LOW]** Empty string `name` bypasses the `None` guard for `addCol`

The validation condition is:
```python
if not transformation_input.col_params or transformation_input.col_params.name is None:
```

`name is None` evaluates to `False` when `name = ""` (empty string). This means a caller can pass an empty column name for `addCol` and bypass the guard. Depending on what `ts.add_column` does with an empty-string name, this could produce a column with a blank header, a runtime error, or silently corrupt the dataframe.

**Suggestion:**
```
Use a stricter truthiness check or explicit length check:
```python
if not transformation_input.col_params or not transformation_input.col_params.name:
    raise HTTPException(status_code=400, detail="Column index and name required for addCol")
```
`not name` is `True` for both `None` and `""`. Alternatively, add a Pydantic `@field_validator` on the `AddOrDeleteColumn` or a dedicated `AddColumn` model to enforce `name` is non-empty when present.
```

File: dataloom-backend/app/schemas.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ class AddOrDeleteRow(BaseModel):
 class AddOrDeleteColumn(BaseModel):
     """Parameters for adding or deleting a column by index and name."""
     index: int
-    name: str
+    name: Optional[str] = None
```
Comment (ivantha) [inline dataloom-backend/app/schemas.py]: **[LOW]** Missing `Optional` import verification

The diff introduces `Optional[str]` in `schemas.py` but does not show the import block. In Python < 3.10, `Optional` must be imported from `typing`:
```python
from typing import Optional
```
If `Optional` was already imported elsewhere in the file this is fine, but it is worth confirming so the change doesn't introduce a `NameError` in environments that haven't run the updated file yet.

**Suggestion:**
```
Verify that `from typing import Optional` (or an equivalent) appears at the top of `app/schemas.py`. In Python 3.10+, the native `str | None` syntax can be used instead:
```python
name: str | None = None
```
```

File: dataloom-backend/app/api/endpoints/transformations.py (hunk 1 above)
Comment (ivantha) [inline dataloom-backend/app/api/endpoints/transformations.py]: **[INFO]** Positive: improved error message specificity

The old error message `"Column parameters required"` was vague and applied to any missing `col_params`. The new message `"Column index and name required for addCol"` correctly scopes the error to the `addCol` operation and names the required fields. This makes API debugging significantly easier for consumers of this endpoint.

**Suggestion:**
```
No change needed — this is a good improvement. Consider applying the same pattern to other operation-specific error messages in the handler for consistency.
```

File: dataloom-frontend/src/Components/Table.jsx [hunk 3]
```diff
@@ -148,9 +148,11 @@ const Table = ({ projectId, data: externalData }) => {
     }
 
     try {
+      const colIndex = index - 1;
+      const colName = columns[index];
       const response = await transformProject(projectId, {
         operation_type: "delCol",
-        col_params: { index: index - 1 },
+        col_params: { index: colIndex, name: colName },
```
Comment (ivantha) [inline dataloom-frontend/src/Components/Table.jsx]: **[INFO]** Positive: extracting intermediate variables improves readability

Extracting `colIndex` and `colName` into named variables before building the request object is a good readability improvement over the inline `index - 1` expression that existed before. It makes the intent of each value clearer and simplifies future debugging.

**Suggestion:**
```
No change needed for the extraction pattern itself — just ensure the values are computed correctly (see the off-by-one finding above).
```

File: dataloom-frontend/src/Components/Table.jsx [hunk 4]
```diff
@@ -148,9 +148,11 @@ const Table = ({ projectId, data: externalData }) => {
     }
 
     try {
+      const colIndex = index - 1;
+      const colName = columns[index];
```
Comment (ivantha) [inline dataloom-frontend/src/Components/Table.jsx:152]: **[HIGH]** Off-by-one error: `columns[index]` retrieves wrong column name

The variable `index` appears to be 1-based (inferred from the fact that `colIndex = index - 1` is used to convert to a 0-based backend index). However, `colName` is computed as `columns[index]` — which, for a 0-based JS array, returns the element at position `index`, i.e., **the next column**, not the one being deleted.

Example: If the user deletes column at UI position 2 (`index = 2`), `colIndex` becomes `1` (correct backend index), but `colName = columns[2]` retrieves the **third** element (index 2) of a 0-based array — the column *after* the one being deleted.

The correct lookup should use `columns[index - 1]` or, equivalently, `columns[colIndex]`.

**Suggestion:**
```
```js
const colIndex = index - 1;
const colName = columns[colIndex]; // was: columns[index]
```
```

File: dataloom-backend/app/api/endpoints/transformations.py (hunk 1 above)
Comment (ivantha) [inline dataloom-backend/app/api/endpoints/transformations.py:61]: **[MEDIUM]** Validation condition is redundant / partially redundant

The new guard `if not transformation_input.col_params or transformation_input.col_params.name is None` has an issue: because `index: int` is still a **required** field in `AddOrDeleteColumn`, Pydantic will already reject a request body that omits `col_params` entirely with a 422 — so the `not transformation_input.col_params` branch should never actually be reached for `addCol`. More importantly, the check does not validate that `col_params.index` is a sane value (e.g., non-negative), which was the intended gap the PR was supposed to improve.

Also, the `index` field on `AddOrDeleteColumn` is still required (`index: int`). If a caller sends `col_params: {}` for `addCol`, Pydantic rejects it before this code runs. The runtime guard only adds value for `name`. Documenting this assumption would make the intent clearer.

**Suggestion:**
```
```python
elif op == 'addCol':
    if not transformation_input.col_params:
        raise HTTPException(status_code=400, detail="Column parameters required for addCol")
    if transformation_input.col_params.name is None:
        raise HTTPException(status_code=400, detail="Column name is required for addCol")
    p = transformation_input.col_params
    return ts.add_column(df, p.index, p.name), True
```
Splitting the checks gives callers clearer error messages and makes each guard's purpose explicit.
```

File: dataloom-frontend/src/Components/Table.jsx [hunk 5]
```diff
@@ -148,9 +148,11 @@ const Table = ({ projectId, data: externalData }) => {
...
       const response = await transformProject(projectId, {
         operation_type: "delCol",
-        col_params: { index: index - 1 },
+        col_params: { index: colIndex, name: colName },
       });
       updateTableData(response);
```
Comment (ivantha) [inline dataloom-frontend/src/Components/Table.jsx:157]: **[MEDIUM]** `name` is sent with `delCol` but appears unused on the backend

The diff adds `name: colName` to the `delCol` payload. However, the backend diff shows no change to the `delCol` handler, and the original 422 was caused purely by Pydantic rejecting a missing required `name` on the schema — not by any backend logic for `delCol` actually needing the name.

Sending `name` to `delCol` is likely harmless today, but it creates a false impression that the backend uses the name for deletion (it presumably uses only the index), leaks unnecessary UI state to the server, and couples the frontend to backend implementation details that don't exist. The cleaner fix was simply making `name` optional in the schema (already done), without adding `name` to the payload.

**Suggestion:**
```
Unless the backend `delCol` handler actually validates or uses `name`, consider reverting the frontend change to keep the payload minimal:
```js
const response = await transformProject(projectId, {
  operation_type: "delCol",
  col_params: { index: index - 1 },
});
```
If the name is intentionally forwarded for audit/logging purposes, add a comment explaining why.
```

File: dataloom-backend/app/api/endpoints/transformations.py [hunk 6]
```diff
@@ -57,8 +57,8 @@ def _handle_basic_transform(df, transformation_input, project, db, project_id):
...
         p = transformation_input.col_params
         return ts.add_column(df, p.index, p.name), True
```
Comment (ivantha) [inline dataloom-backend/app/api/endpoints/transformations.py:63]: **[MEDIUM]** No automated tests added for changed validation logic

The validation path for `addCol` was changed (new condition, new error message) and the schema was relaxed. Despite these behavioral changes, no unit or integration tests were added. The PR checklist confirms "New tests added" is unchecked. Without tests:
- A future refactor could silently re-introduce the 422 for `delCol`.
- The new condition `col_params.name is None` is untested at the unit level.
- Edge cases (e.g., `addCol` with `name=""`, `name=" "`, `col_params` present but `name` absent) are unverified.

**Suggestion:**
```
Add pytest cases covering at minimum:
1. `delCol` with no `name` → 200 OK
2. `addCol` with `name` provided → 200 OK
3. `addCol` with `name=None` → 400 with message "Column index and name required for addCol"
4. `addCol` with `col_params` missing → 422 from Pydantic
```

File: dataloom-backend/app/schemas.py (hunk 2 above)
Comment (ivantha) [inline dataloom-backend/app/schemas.py:96]: **[LOW]** Class docstring is stale after making `name` optional

The `AddOrDeleteColumn` docstring still reads *"Parameters for adding or deleting a column by index and name."* After this PR, `name` is optional and only required for additions. The docstring misleads readers into thinking `name` is always required.

**Suggestion:**
```
```python
class AddOrDeleteColumn(BaseModel):
    """Parameters for adding or deleting a column by index.
    `name` is required for addCol operations and optional for delCol."""
    index: int
    name: Optional[str] = None
```
```

File: dataloom-backend/app/api/endpoints/transformations.py (hunk 1 above)
Comment (ivantha) [inline dataloom-backend/app/api/endpoints/transformations.py:61]: **[LOW]** Error message change silently alters API contract

The error detail string changed from `"Column parameters required"` to `"Column index and name required for addCol"`. If any downstream client (tests, scripts, monitoring alerts) matches the old error string, they will silently break. The change itself is an improvement in clarity, but it should be noted as a minor behavioral change in the PR.

**Suggestion:**
```
Document the message change in the PR description. If the project has tests that assert on the exact error string, update them. Consider using a constant for error messages to avoid future drift:
```python
COL_PARAMS_REQUIRED_MSG = "Column index and name required for addCol"
```
```

File: dataloom-frontend/src/Components/Table.jsx (hunk 4 above)
Comment (ivantha) [inline dataloom-frontend/src/Components/Table.jsx:152]: **[INFO]** Intermediate variables add clarity but naming could be more explicit

Extracting `colIndex` and `colName` into named variables before building the payload is a good readability improvement over the inline `index - 1` expression. However, `colName` could be named `columnName` for consistency with the rest of the codebase style, and a comment explaining why 1 is subtracted would help future readers.

**Suggestion:**
```
```js
const columnIndex = index - 1; // convert from 1-based UI index to 0-based backend index
const columnName = columns[columnIndex];
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR correctly identifies and addresses a real 422 error caused by Pydantic rejecting `delCol` requests that lacked the previously-required `name` field. The schema relaxation (`name: Optional[str] = None`) and the added `addCol` guard are sound in intent. However, there is a likely off-by-one bug in `Table.jsx` where `columns[index]` should be `columns[index - 1]` (or `columns[colIndex]`) — causing the wrong column name to be looked up when `index` is 1-based. The PR description also contains a direct factual error: it claims the frontend change *removes* the name parameter, when it actually *adds* it. No automated tests were added to cover the changed validation paths, and the PR template has unfilled placeholders (issue number, testing checklist) that should be addressed before merging.

**Suggested Version Bump:** `patch`

### ⚠️ Breaking Changes

- **[INFO]** Schema relaxation: `AddOrDeleteColumn.name` changed from required to optional: **Previous behavior:** Any request to `/transformations` endpoints that included `col_params` was required to supply a non-null `name` string. Omitting it caused Pydantic to return a 422.

**New behavior:** `name` is `Optional[str] = None`. Requests without `name` are now accepted by the schema layer.

**Who is affected:** All callers of transformation endpoints that use `col_params`. Any caller that was *already* supplying `name` is unaffected. Callers that relied on the 422 for input validation (e.g., integration tests asserting a 422 for missing `name`) will now get a different response.

**Migration path:** No changes needed for existing callers sending `name`. Tests asserting a 422 for missing `name` must be updated to expect either a 200 (for `delCol`) or a 400 (for `addCol`, now enforced in application code).
  > **Suggestion:** This is a backward-compatible relaxation (existing valid payloads continue to work). Suggest **patch** semver bump.
- **[LOW]** Error message text change in `addCol` validation path: **Previous behavior:** Calling `addCol` without `col_params` returned HTTP 400 with `detail: "Column parameters required"`.

**New behavior:** The same scenario now returns HTTP 400 with `detail: "Column index and name required for addCol"`.

**Who is affected:** Any client, test, or monitoring rule that pattern-matches on the exact error detail string.

**Migration path:** Update any client code or test assertions that compare against the old error string `"Column parameters required"`.
  > **Suggestion:** Use a named constant for the error detail string to make future changes explicit and searchable:
```python
ADD_COL_VALIDATION_ERROR = "Column index and name required for addCol"
```


### PR Description Issues

- **[HIGH]** PR description contradicts the actual code change (description-diff mismatch): The **"Changes Made / Frontend"** section states:
> *"Updated the transformProject call in Table.jsx to **remove** the redundant name parameter during column deletion."*

However, the actual diff does the **opposite**: it **adds** `name: colName` to the `delCol` payload. This is a direct factual inaccuracy in the description and will confuse reviewers and future git historians trying to understand the intent of the commit.
  > **Suggestion:** Replace the misleading description with an accurate one, e.g.:
> **Frontend:** Updated `Table.jsx` to capture the column name from the `columns` array and pass it alongside the index in the `delCol` payload. This aligns with the now-optional `name` field in the schema and provides the backend with richer context for future use.
- **[MEDIUM]** Missing issue link and incomplete testing section: The PR description includes `Fixes #(issue number)` as a literal template placeholder — the issue number was never filled in. Additionally, the "How Has This Been Tested?" section only has manual testing checked; there is no mention of which existing test suites were run (the checklist says "Existing tests pass" is unchecked), even though the PR claims all tests passed locally.
  > **Suggestion:** 1. Replace `Fixes #(issue number)` with the actual issue reference (e.g., `Fixes #142`) or remove the line if there is no linked issue.
2. Check the "Existing tests pass" checkbox and describe which test suites were run (e.g., `pytest`, `vitest`/`jest`) and their results.
- **[LOW]** "Breaking Change" section left empty without explicit confirmation: The PR template includes a "Breaking change" checkbox which is unchecked and no breaking-change section is provided. This PR does relax the API schema (`name` from required to optional) and changes an error message string — both of which could be considered behavioral changes. A brief note confirming the PR author's assessment would improve transparency.
  > **Suggestion:** Add a short note under a **Breaking Changes** heading:
> *None. The `name` field in `AddOrDeleteColumn` was made optional (additive/relaxing change). The `addCol` error message text changed from "Column parameters required" to "Column index and name required for addCol" — clients matching the exact string should update their error handling.*

**Comment by ivantha** [comment]
Thanks for the PR! It currently has merge conflicts with main. Could you rebase onto the latest main and push the updated branch? Happy to merge once it's auto-rebasable.

**Comment by ivantha** [comment]
Thank you for tracking down the 422 error on column deletion! Unfortunately this PR conflicts with current main. Please rebase and reopen — the bug fix is valuable.
